pkg/resource/deploy/step.go

```diff
@@ -102,6 +102,7 @@ func NewCreateStep(plan *Plan, reg RegisterResourceEvent, new *resource.State) S
 	contract.Assert(new.URN != "")
 	contract.Assert(new.ID == "")
 	contract.Assert(!new.Delete)
+	contract.Assert(!new.External)
```

Comment:
This just made me think of an alternative design. I'm not advocating it per se, however it could be simpler.

First, recall that we are likely to want to do a refresh operation before performing updates. See https://github.com/pulumi/pulumi-terraform/issues/57. As soon as we do that, we are going to be doing more "read"-like operations. It's worth considering how we will model those in the engine, given these changes.

Second, it occurs to me that we could actually just say that, instead of `OpRead` and `OpReadReplacement`, the existing operations `OpCreate`, `OpUpdate`, and `OpCreateReplacement` handle these functions. They would simply mean something different when `External` is `true`, just as `OpDelete` is different when `External` is `true`. The nice thing is that this may help inch us towards https://github.com/pulumi/pulumi-terraform/issues/57.

This is appealing to me because on the UI, it's not clear to me we want to expose these operations so prominently. So we'd just keep our existing visualizations and merely present them differently for external resources in a consistent way. I'm also sure *some* people will want to see them, but because they are external to the stack, it's probably far less interesting.  And, indeed, it may be alarming, since most people assume resources fetched in this manner are *not* under the management of Pulumi. If they see one getting deleted, they are likely to freak out. So I suspect we'd want to filter them out by default or summarize them in a specific way.

Did you consider this approach? I'm curious what you think about the tradeoffs involved?

Reply:
@swgillespie and I discussed this approach a bit last week. Personally, I am in favor of something like what you describe: my gut says that the resulting code will be a bit simpler. We agreed to go ahead with the current strategy, though, and make a final decision after seeing how things turned out.

Reply:
Pat had the same (or a similar) idea when we whiteboarded this last week. Here are my thoughts:

It is possible to implement refresh in terms of both `OpRead` and `OpCreate` and friends. In both cases, however, it is difficult to implement https://github.com/pulumi/pulumi-terraform/issues/57, because the step generator must know the results of the refresh before deciding what kind of step to generate for the resource itself. This is in contrast to the step generator's normal mode of operation, where steps are generated using only the knowledge of incoming new inputs and existing old inputs. Therefore, this implementation strategy is less of a question about implementation efficiency and correctness but more of a question of software engineering, to steal Pat's quote.

I lean towards having small, composable steps whose semantics are small over larger steps whose semantics are more complex. The reason for this is that it pushes complexity away from step application and towards the step generator, where we can centralize this logic in one place that can be tested as a single unit. If the outputs of the step generator (steps) are small and have well-known and simple semantics, in my mind it is easier to reason about the correctness of a plan as a whole because it is clear just by looking at the steps themselves (and not their inputs) what a particular step will do.

Having `OpCreate`, `OpUpdate`, `OpCreateReplacement` change semantics depending on whether or not their target(s) are `External` flies a little bit in the face of that tendency I have. When looking at a series of steps generated by the step generator, one would also have to know the target resources (whether or not they were `External`) to fully understand the semantics of the step.

That said, the fact that `OpDelete` handles both `External` and non-`External` resources in this PR is a contradiction to everything I have just said. I actually did have a `read-delete` step in a previous version of this PR. Here's how I justify it, though: the primitives of Pulumi are read, update, create, and delete. All operations can be phrased in terms of those four operations. In my mind it makes sense to conflate `OpDelete` with the deletion of an external resource, because conceptually we're still *deleting* - we're just removing something from a snapshot. However, in my mind it does *not* make sense to conflate `read` and `create`, since fundamentally those are two very different operations that bottom out at different provider operations.